KEP 1287: Graduate InPlacePodVerticalScaling to GA

natasha41575 opened the pull request:
<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Graduate InPlacePodVerticalScaling to GA.

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/1287

<!-- other comments or additional information -->
- Other comments: We believe we are ready to graduate InPlacePodVerticalScaling to GA in kubernetes 1.35. See comments below for how we have met the graduation criteria, and more details in the KEP for other updates.

/sig node autoscaling scheduling

/cc @tallclair @dchen1107

/cc @dom4ha 
for sig-scheduling

/cc @jackfrancis 
for sig-autoscaling

natasha41575 commented:
/cc @wojtek-t

jackfrancis commented on keps/sig-node/1287-in-place-update-pod-resources/README.md:
```diff
@@ -467,6 +478,12 @@ Allocation will be attempted on the pods in the queue:
 A successful allocation will trigger a pod sync, which will actuate the allocated resize and update the
 pod status accordingly.
 
+### Kubelet-triggered eviction
```
Should we add this newly described (to this KEP) eviction behavior to the implementation history so it doesn't seem like we are simultaneously adding new functionality at the same time that we're prepping for GA graduation?

jackfrancis commented on keps/sig-node/1287-in-place-update-pod-resources/README.md:
```diff
@@ -984,18 +997,15 @@ later accepted. This metric primarily exists because if a deferred resize is acc
 Labels:
-  - `accepted_reason` - whether the resize was accepted through the timed retry or due to another pod event. Possible values: `periodic_retry`, `event_based`.
-  - `namespace` - the namespace of the pod.
+  - `retry_trigger` - whether the resize was accepted through the timed retry or due to another pod event. Possible values: `periodic_retry`, `pod_resized`, `pod_updated`, `pods_added`, `pods_removed`.
```
ditto my above comments about implementation history to explain the evolution of the labels in the `kubelet_pod_deferred_resize_accepted_total` metric

natasha41575 replied:
same response as above

tallclair reviewed (commented):
/lgtm

jackfrancis commented on keps/sig-node/1287-in-place-update-pod-resources/README.md:
```diff
@@ -934,47 +947,47 @@ Labels:
 
+This metric is recorded as a counter.
+
+#### `kubelet_pod_pending_resizes`
```
thank you!

jackfrancis reviewed (approved):
/lgtm

for SIG Autoscaling

cc @adrianmoisey

haircommander commented on keps/sig-node/1287-in-place-update-pod-resources/README.md:
```diff
@@ -811,7 +824,7 @@ be a race condition where the Kubelet may or may not accept the first resize, de
 intended.
 
-The atomic resize requirement should be reevaluated prior to GA, and in the context of pod-level resources.
+The atomic resize requirement should be reevaluated in the context of pod-level resources.
```
since this isn't a requirement here, maybe just say `may be reevaluated` ? like I don't think that would block pod level resources either, so probably we do it as a separate kep from either this or that

haircommander commented:
one note, but it can be done as a follow up
/lgtm

from sig node

SergeyKanzhelev commented:
/hold

@natasha41575 can you please do more changes

## Stable criteria:

Comment on the PR on all the GA criteria and clarify if they are satisfied:

-  VPA integration of feature moved to beta,
- User feedback (ideally from at least two distinct users) is green,
- No major bugs reported for three months.
- Pod-scoped resources are handled if that KEP is past alpha
- UpdatePodSandboxResources is implemented by containerd & CRI-O
- Re-evaluate the following decisions:
    - Resize atomicity
    - Exposing allocated resources in the pod status
    - QOS class changes

For example, containerd seems to be stalling on implementing the update methos: https://github.com/containerd/containerd/pull/11406/.

Criteria that were decides to be postponed needs to be removed from the KEP, but we need to be very explicit why they are moved.

## Conformance

Please identify the set of tests that will be implemented as conformance. We are trying to be more strict on all ensuring all new APIs are covered with the Conformance tests.

## Tests

There are still a `TODO: Identify more cases` in tests section. This needs to be cleaned up.

I think the coverage is lacking all the failure cases when CRI API is not accepting the change. These tests can be implemented with the CRI proxy we now have in tests.

Another coverage needs to be added to swap tests to ensure it is not working with inplaceresize as stated in the KEP.

SergeyKanzhelev commented:
/cc @chrishenzie 

for Containerd and `UpdatePodSandboxResources`

haircommander commented:
/lgtm cancel

SergeyKanzhelev commented:
More specific questions:

For VPA I see the FG is in beta: https://github.com/kubernetes/autoscaler/blob/d789bff8b754a6d05ff62c223b7bf26df09b2c30/vertical-pod-autoscaler/pkg/features/features.go#L50, but I wonder if we want:
1. Get confirmation that InPlaceOnly is feasible to implement and whether we want this support be in beta. 
2. Do we want VPA to support sidecar resize since this KEP implemented it later and we need some confirmation on it.

User feedback and recent bugs - I think we are OK on those as far as I know and made a quick check.

Pod-scoped resources are handled if that KEP is past alpha - the KEP is passed alpha. So we need to change this criteria.

For "Re-evaluate the following decisions" I suggest we actually write up the re-evaluation and a final decision.

natasha41575 commented on keps/sig-node/1287-in-place-update-pod-resources/README.md:
```diff
@@ -1305,15 +1389,30 @@ TODO: Identify more cases
-- No major bugs reported for three months.
-- Pod-scoped resources are handled if that KEP is past alpha
-- `UpdatePodSandboxResources` is implemented by containerd & CRI-O
+- VPA integration of feature, `InPlaceOrRecreate` update mode, is moved to beta
```
This has moved to beta

natasha41575 commented:
> /hold
> 
> @natasha41575 can you please do more changes
> 
> ## Stable criteria:
> Comment on the PR on all the GA criteria and clarify if they are satisfied:
> 
> * VPA integration of feature moved to beta,
> * User feedback (ideally from at least two distinct users) is green,
> * No major bugs reported for three months.
> * Pod-scoped resources are handled if that KEP is past alpha
> * UpdatePodSandboxResources is implemented by containerd & CRI-O
> * Re-evaluate the following decisions:
>   
>   * Resize atomicity
>   * Exposing allocated resources in the pod status
>   * QOS class changes
> 
> For example, containerd seems to be stalling on implementing the update methos: [containerd/containerd#11406](https://github.com/containerd/containerd/pull/11406).

Done.  Yes, I'm aware of the containerd issues. It's slated to be released in containerd 2.2 in November, but there won't be enough time to provide substantial test coverage of it. I had a chat with @tallclair though and we agreed that this doesn't need to be blocking. IMO it's a very small and noncritical part of the feature. 

> Criteria that were decides to be postponed needs to be removed from the KEP, but we need to be very explicit why they are moved.

I added a paragraph below the stable criteria in the KEP to explain what was removed and why. 

> ## Conformance
> Please identify the set of tests that will be implemented as conformance. We are trying to be more strict on all ensuring all new APIs are covered with the Conformance tests.

Added a section under the stable criteria to point to which specific tests will be moved to Conformance. 

> ## Tests
> There are still a `TODO: Identify more cases` in tests section. This needs to be cleaned up.

Done. 

> I think the coverage is lacking all the failure cases when CRI API is not accepting the change. These tests can be implemented with the CRI proxy we now have in tests.

Yeah, we will add tests to cover the CRI API when we can but again I don't think we do that prior to code freeze, since containerd 2.2 won't be available until November (and I imagine it takes some time for test infra to pick up the newest containerd?) 

> Another coverage needs to be added to swap tests to ensure it is not working with inplaceresize as stated in the KEP.

Swap has substantial coverage in the unit tests here: https://github.com/kubernetes/kubernetes/blob/ad82c3d39f5e9f21e173ffeb8aa57953a0da4601/pkg/kubelet/allocation/allocation_manager_test.go#L910

I feel confident enough in these unit tests that I don't think we need to add e2e coverage for swap specifically, but let me know if you have a different opinion

natasha41575 commented:
> More specific questions:
> 
> For VPA I see the FG is in beta: https://github.com/kubernetes/autoscaler/blob/d789bff8b754a6d05ff62c223b7bf26df09b2c30/vertical-pod-autoscaler/pkg/features/features.go#L50, but I wonder if we want:
> 
> 1. Get confirmation that InPlaceOnly is feasible to implement and whether we want this support be in beta.
> 2. Do we want VPA to support sidecar resize since this KEP implemented it later and we need some confirmation on it.

I think the answers from VPA owners above have answered these questions. 

> User feedback and recent bugs - I think we are OK on those as far as I know and made a quick check.

+1 

> Pod-scoped resources are handled if that KEP is past alpha - the KEP is passed alpha. So we need to change this criteria.

Yeah I added a note about this to the graduation criteria. Basically yes it is now beta. Support for IPPR for pod-level resources is slated to land in 1.35 (cc @ndixita). I think it is really important for that to land but I wouldn't consider it strictly blocking. I am of the opinion that it should block GA of pod-level resources, but that is a different discussion.

> For "Re-evaluate the following decisions" I suggest we actually write up the re-evaluation and a final decision.

Added to the KEP.

jackfrancis commented:
@adrianmoisey let me know if you're unable to do follow-up review for SIG Autoscaling, I'd be happy to do so but would like you to have the first shot as VPA stakeholder

maxcao13 commented:
> Get confirmation that InPlaceOnly is feasible to implement and whether we want this support be in beta.

Just asking for clarification, we are talking about the feasibility to implement an additional `InPlaceOnly` update mode from the VPA perspective (as opposed to the currently implemented mode `InPlaceOrRecreate`), and whether this should be criteria for InPlacePodVerticalScaling to GA?

If so, my opinion is that it shouldn't be a blocker. Because in my mind, `InPlaceOnly` would just be a subset of `InPlaceOrRecreate` in that we simply just _do nothing_, when we can't do an in place resize vs falling back to `Recreate`. So we are already covering the use case of getting VPA users to use the new feature.

SergeyKanzhelev commented:
> > > Get confirmation that InPlaceOnly is feasible to implement and whether we want this support be in beta.
> > 
> > 
> > It's feasible to implement but I don't think we want this support in beta ( it should be an alpha feature first ), maybe @adrianmoisey thinks differently..
> 
> Agreed, if we add InPlaceOnly to VPA, that's a new feature that will start from alpha.

The question here is whether the validation of this KEP is satisfactory to the spirit of the graduation criteria we agreed on before. The KEP was originally planned as a way for VPA to offer quick non-disruptive workload right-sizing. Original implementation didn't satisfy the VPA as it did not guarantee non-disruptive part and didn't offer a way to neither predict disruption nor ask to not disrupt, nor even understand that disruption was caused by resize. So in beta API was changed to improve the promise of non-disruption.

Meanwhile in VPA the only mode implemented as I understand is disruptive. And the desired non-disruptive mode is not even in alpha.

Now the question is weather VPA customers have enough value from InPlaceOrRestart that it will live for a long time with it. If not and the desired state is to have InPlaceOnly, than the graduation criteria was asking to get VPA implementation to be at least beta for customers adoption.

Again, the spirit of this graduation criteria was to get VPA confirmation that what was the original goal of this KEP is achieved and validated. I do not see how this is satisfied. However if we change a goal (we see that InPlaceOnly not needed for VPA customers) or we believe that InPlaceOrRestart gave enough validation for InPlaceOnly, please confirm

SergeyKanzhelev commented:
my objective is to make sure we are on the same page on what is implemented and what VPA uses.

> Understood. But from my understanding, InPlacePodVerticalScaling will never be fully disruption free (current resize policies of `NotRequired` and `RestartContainer` and the absence of a `GuaranteedNoRestart` or something), and I believe that's why we on the VPA side relaxed our position of promising a disruption free resize mode.

Yes, this is true. However `NotRequired` offers the "best effort" for not restarting with the exact semantic of "we have no intention to break it by resize". I think the future direction will be to minimize the number of "misses" when we actually do break apps.
 
> I'll defer to the maintainers on this, but that was my impression during the development of the VPA implementation.

Agree to get an opinion here. If the goal is changed to InPlaceOrRestart, it is also fine. It is just not what we discussed before.
 
> Unless I am misunderstanding, there is not way for an /resize API call to guarantee that a resize will be disruptionless, where we could just get an API response of "we can't do a resize without restarting the container, so instead we will just do nothing and return this error". If that were the case, then I think we could definitely have an `InPlaceOnly` mode that would satisfy the VPA's original goal.

This is how KEP is written now:

  >> Note: NotRequired restart policy for resize does not guarantee that a container won't be restarted. If the runtime knows a resize will trigger a restart, it should return an error instead, and the Kubelet will retry the resize on the next pod sync. The behavior when shrinking memory limits is defined under [Memory Limit Decreases](https://github.com/kubernetes/enhancements/blob/60374d0ea8f7eac856c132bc952d2fdbde29f4c3/keps/sig-node/1287-in-place-update-pod-resources/README.md#memory-limit-decreases) below.

So I wonder if this is aligned with what you wrote here. If the result of "failed" attempt is that kubelet keeps retrying - will VPA somehow detect and cancel it? Or implementation is different from what is written in KEP and the error will actually bubble up to the VPA.

One possibility here is that VPA and customer feedback is all from containerd where all resizes are allowed and reasonably safe. So we have a skewed feedback on this. So all the improvements made to handle rejection cases to help VPA to work universally in different environments were not validated. If this is the case, is it a release blocker? What kind of validation we can get from VPA that the API provided for resize rejections is satisfactory for VPA?

natasha41575 commented:
Most of @SergeyKanzhelev's comment should probably be answered by VPA folks but will chime in on this part:

> So I wonder if this is aligned with what you wrote. If the result of "failed" attempt is that kubelet keeps retrying - will VPA somehow detect and cancel it? Do you see errors bubble up to VPA in case of InPlaceOrRestart? Or implementation is different from what is written in KEP and the error will actually bubble up to the VPA.

There are multiple failure modes, and VPA handles all of them. These are described here: https://github.com/kubernetes/autoscaler/blob/9c4219866a07bba5adb678c9e5b85628191112d2/vertical-pod-autoscaler/enhancements/4016-in-place-updates-support/README.md?plain=1#L169

Both what is written in the KEP is accurate (kubelet will keep retrying) and the error gets bubbled up to VPA. In the worst case, if the resize hasn't completed by some set amount of time, it will consider it failed and handle it accordingly.

adrianmoisey commented:
Heya,

I've read this thread, and I'm trying to figure out exactly what is needed. Sorry if I missed it, but it seems to be:

> Again, the spirit of this graduation criteria was to get VPA confirmation that what was the original goal of this KEP is achieved and validated

This may be a tough question to answer. Currently the only active VPA maintainers are very new, and weren't around with this KEP was originally created.
I'm unsure what the spirit of the previous discussions were, so it's hard to answer the question of "did we meet the goal".

I assume [that this is the goal](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/1287-in-place-update-pod-resources/README.md#goals) that you're talking about:

> Primary: allow to change container resource requests & limits without necessarily restarting the container.
> Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide how to proceed if in-place resource resize is not possible.
> Secondary: allow users to specify which Containers can be resized without a restart.

For the first point: Yes, I think we hit it. The key word here is "without necessarily restarting the container".
For the second point: Yes, we hit that. The VPA has taken a decision to recreate the Pod if a resize is infeasible

dchen1107 commented:
/lgtm
/approve
